core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -31,256 +31,216 @@
-    super(bufferAllocator, maxMessageSize);
+  protected interface Sink {
+    /**
+     * Sends response headers to the remote end point.
+     *
+     * @param headers the headers to be sent to client.
+     */
+    void writeHeaders(Metadata headers);
+
+    /**
+     * Sends an outbound frame to the remote end point.
+     *
+     * @param frame a buffer containing the chunk of data to be sent.
+     * @param flush {@code true} if more data may not be arriving soon
+     */
+    void writeFrame(@Nullable WritableBuffer frame, boolean flush);
+
+    /**
+     * Sends trailers to the remote end point. This call implies end of stream.
+     *
+     * @param trailers metadata to be sent to the end point
+     * @param headersSent {@code true} if response headers have already been sent.
+     */
+    void writeTrailers(Metadata trailers, boolean headersSent);
+
+    /**
+     * Requests up to the given number of messages from the call to be delivered. This should end up
+     * triggering {@link TransportState#requestMessagesFromDeframer(int)} on the transport thread.
+     */
+    void request(int numMessages);
+
+    /**
+     * Tears down the stream, typically in the event of a timeout. This method may be called
+     * multiple times and from any thread.
+     *
+     * <p>This is a clone of {@link ServerStream#cancel()}.
+     */
+    void cancel(Status status);
+  }
+
+  private final MessageFramer framer;
+  private boolean outboundClosed;
+  private boolean headersSent;
+
+  protected AbstractServerStream(WritableBufferAllocator bufferAllocator) {
+    framer = new MessageFramer(this, bufferAllocator);
   }
 
+  @Override
+  protected abstract TransportState transportState();
+
   /**
-   * Sets the listener to receive notifications. Must be called in the context of the transport
-   * thread.
+   * Sink for transport to be called to perform outbound operations. Each stream must have its own
+   * unique sink.
    */
-  public final void setListener(ServerStreamListener listener) {
-    this.listener = checkNotNull(listener);
-
-    // Now that the stream has actually been initialized, call the listener's onReady callback if
-    // appropriate.
-    onStreamAllocated();
-  }
+  protected abstract Sink abstractServerStreamSink();
 
   @Override
-  protected ServerStreamListener listener() {
-    return listener;
+  protected final MessageFramer framer() {
+    return framer;
   }
 
   @Override
-  protected void receiveMessage(InputStream is) {
-    inboundPhase(Phase.MESSAGE);
-    listener().messageRead(is);
+  public final void request(int numMessages) {
+    abstractServerStreamSink().request(numMessages);
   }
 
   @Override
   public final void writeHeaders(Metadata headers) {
     Preconditions.checkNotNull(headers, "headers");
 
-    outboundPhase(Phase.HEADERS);
     headersSent = true;
-    internalSendHeaders(headers);
-    outboundPhase(Phase.MESSAGE);
+    abstractServerStreamSink().writeHeaders(headers);
```

Comment:
[ServerCallImpl is already doing state checking](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ServerCallImpl.java#L105). And I trust `ServerCallImpl` a _lot_ more than I trust the streams doing that checking.